webdictionary/src/main/java/com/eis0/kademlianetwork/informationdeliverymanager/IdFinderHandler.java

```diff
@@ -0,0 +1,184 @@
+ * to avoid the direct erasure.
+ * <p>
+ * When a node doesn't respond its staleCount get incremented, in the Refresh phase it will be
+ * removed if still inactive.
+ *
+ * @author Marco Cognolato (created the class)
+ * @author Enrico Cestaro (edited and restored the code)
+ * @author Edoardo Raimondi (added the check cicle to verify if the target node is 'alive')
+ */
+public class IdFinderHandler {
+    private static int maxAttempts = 10;
+    private static final String ID_TO_FIND_NULL = "The idToFind cannot be null";
+    private static final String SEARCHER_NULL = "The searcher cannot be null";
+    private static final String NODES_LIST_NULL = "The nodeList cannot be null";
+    private static final String NODES_LIST_EMPTY = "The nodeList cannot be empty";
+private static final String INSUFFICIENT_ATTEMPTS = "The attemptsNumber must be higher then zero";
+
+    /**
+     * Searches for a specific Id, which needs to be closer to a given resource Id.
+     * The list of possible nodes is formed by the closest nodes, and it always contains the local
+     * node (if it doesn't, the node is automatically added)
+     *
+     * @param idToFind The ID to find in the network
+     * @param searcher The Peer which is searching for the ID
+     * @throws NullPointerException If the idToFind, the searcher or the researchMode are null
+     */
+    public static void searchId(@NonNull KademliaId idToFind, @NonNull SMSPeer searcher) {
+        if (idToFind == null) throw new NullPointerException(ID_TO_FIND_NULL);
+        if (searcher == null) throw new NullPointerException(SEARCHER_NULL);
+        //Obtain the local RoutingTable
+        SMSKademliaRoutingTable table = KademliaJoinableNetwork.getInstance().getLocalRoutingTable();
+        //Create a list containing the closest nodes (five is more than enough)
+        List<SMSKademliaNode> nodeList = table.findClosest(idToFind, maxAttempts);
+
+        searchIdList(idToFind, searcher, nodeList);
+    }
+
+    /**
+     * Searches inside of the given list for the active closest node; when a node doesn't respond it
+     * gets removed from the list, and the process is repeated
+     *
+     * @param idToFind The ID to find in the network
+     * @param searcher The peer which is searching for the ID
+     * @param nodeList
+     */
+    public static void searchIdList(@NonNull KademliaId idToFind, @NonNull SMSPeer searcher,
+                                    @NonNull List<SMSKademliaNode> nodeList) {
+        if (nodeList.isEmpty()) throw new IllegalArgumentException(NODES_LIST_EMPTY);
+        if (nodeList == null) throw new NullPointerException(NODES_LIST_NULL);
+
+        //Local node, inserted in the list if not already present
+        SMSKademliaNode local = KademliaJoinableNetwork.getInstance().getLocalNode();
+        if (!nodeList.contains(local)) {
+            //inserted in the last position, must be the last node to be checked
+            nodeList.add(local);
+        }
```

Comment:
Here if in the list you add the local node, but later (on `retryIfDead`) you remove it. Even though the local node will definitely answer, since the message is routed to himself, I still think it *might* be possible to have an infinite cicle where you remove it, then you add it, then you remove it, etc.
I'd suggest to move this in a setup place, where you add it once, (so inside `searchId`), just to be *extra* sure.

Reply:
I remember moving it from searchId to searchIdList for a reason, but it may have been for a mistake that I have later corrected; anyway you are right